tls/x509_crt.c

```diff
@@ -924,76 +924,58 @@ ttls_x509_crt_parse(TlsX509Crt *chain, unsigned char *buf, size_t buflen)
 	}
 
 	if (buf_format == TTLS_X509_FORMAT_DER)
-		return ttls_x509_crt_parse_der(chain, buf, buflen);
-
-	if (buf_format == TTLS_X509_FORMAT_PEM) {
-		/*
-		 * 1 rather than 0 since the terminating NULL byte
-		 * is counted in.
-		 */
-		while (buflen > 1) {
-			int r;
-			size_t use_len;
-			unsigned char *pem_dec;
+		if ((r = ttls_x509_crt_parse_der(crt, buf, buflen)))
+			goto done;
 
-			/*
-			 * If we get there, we know the string is
-			 * null-terminated.
-			 */
-			r = ttls_pem_read_buffer("-----BEGIN CERTIFICATE-----",
-						 "-----END CERTIFICATE-----",
-						 buf, &use_len);
-			if (r > 0) {
-				/* Was PEM encoded. */
-				pem_dec = buf;
-				buflen -= use_len;
-				buf += use_len;
-			}
-			else if (r != TTLS_ERR_PEM_NO_HEADER_FOOTER_PRESENT) {
-				/* PEM header and footer were found. */
-				buflen -= use_len;
-				buf += use_len;
-
-				if (!first_error)
-					first_error = r;
-				total_failed++;
-				continue;
-			}
-			else {
-				break;
-			}
+	if (buf_format != TTLS_X509_FORMAT_PEM)
+		return TTLS_ERR_X509_CERT_UNKNOWN_FORMAT;
 
-			r = ttls_x509_crt_parse_der(chain, pem_dec, r);
-			if (r) {
-				/* Quit parsing on a memory error. */
-				if (r == TTLS_ERR_X509_ALLOC_FAILED)
-					return r;
-				if (!first_error)
-					first_error = r;
-				total_failed++;
-				continue;
-			}
+	/*
+	 * buflen > 1 rather than 0 since the terminating NULL byte
+	 * is counted in.
+	 */
+	for ( ; buflen > 1 && crt_len_len; crt_len_len -= TTLS_CERT_LEN_LEN) {
+		size_t use_len;
+		unsigned char *pem_dec;
 
-			success = 1;
+		/*
+		 * If we get there, we know the string is
+		 * null-terminated.
+		 */
+		r = ttls_pem_read_buffer("-----BEGIN CERTIFICATE-----",
+					 "-----END CERTIFICATE-----",
+					 buf, &use_len);
+		if (r > 0) {
+			/* Was PEM encoded. */
+			pem_dec = buf;
+			buflen -= use_len;
+			buf += use_len;
+		}
+		else if (r != TTLS_ERR_PEM_NO_HEADER_FOOTER_PRESENT) {
```

Comment:
`r`can be `TTLS_ERR_PEM_FEATURE_UNAVAILABLE` and `TTLS_ERR_PEM_INVALID_DATA` or `TTLS_ERR_PEM_INVALID_DATA + ret`. why we continue parsing certificates here if error occurs?

Reply:
I check it is error here. If we create certificate like this
```
-----BEGIN CERTIFICATE-----
MIIBkTCCARgCAQEwCgYIKoZIzj0EAwMwRTELMAkGA1UEBhMCQVUxEzARBgNVBAgM
ClNvbWUtU3RhdGUxITAfBgNVBAoMGEludGVybmV0IFdpZGdpdHMgUHR5IEx0ZDAe
Fw0yNDA3MDQxMTI0NDZaFw0yNTA3MDQxMTI0NDZaMEUxCzAJBgNVBAYTAkFVMRMw
EQYDVQQIDApTb21lLVN0YXRlMSEwHwYDVQQKDBhJbnRlcm5ldCBXaWRnaXRzIFB0
eSBMdGQwWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAASdAgJZTIzvmeLYyWEPAt6D
xsALWO8iLk54JMA3UwNi72lAvTw6VVPiH6K5lESlPAoOayge4oYEp6v8s409J5Z/
MAoGCCqGSM49BAMDA2cAMGQCMGT+Gr+sibbbUI5TRhhMcg+UgSAPVizAjmxpWdtj
RVSsv+6Gh+LxvwqCoQW8vWJOPQIwK+08D7up/Ysguy5v8z+iTrfNhjUnqie0z1LQ
Ryep+3L/W0aw+ySuSI0paUxeNN5K
-----END CERTIFICATE-----
-----BEGIN CERTIFICATE-----
Proc-Type: 4,ENCRYPTEDNwYq7pZxPSuAtKbGmi/0dEoHswCgYIKoZIzj0EAwMw
RTELMAkGA1UEBhMCQVUxEzARBgNVBAgMClNvbWUtU3RhdGUxITAfBgNVBAoMGElu
dGVybmV0IFdpZGdpdHMgUHR5IEx0ZDAeFw0yNDA3MDQxMTIzMzNaFw0yNDA4MDMx
MTIzMzNaMEUxCzAJBgNVBAYTAkFVMRMwEQYDVQQIDApTb21lLVN0YXRlMSEwHwYD
VQQKDBhJbnRlcm5ldCBXaWRnaXRzIFB0eSBMdGQwdjAQBgcqhkjOPQIBBgUrgQQA
IgNiAASaXikki3uhfLWfzcIb6A02bAFasl9nyPrvSnggLThRi/JSVDnA11Gtvsfs
zwB9TQyXPwvp/yNd6HfxdsklAhLX5ekx1MiMd23LoMKegmdexUB/PW/SrA5sPeRX
CWk0MPSjUzBRMB0GA1UdDgQWBBS23OcvU3qYQiHsQUueylpPSBKwLjAfBgNVHSME
GDAWgBS23OcvU3qYQiHsQUueylpPSBKwLjAPBgNVHRMBAf8EBTADAQH/MAoGCCqG
SM49BAMDA2cAMGQCMA351Zabc8zAAlVhCjG1bEziJ92t80MfaoNePWBcpNycBoBd
EDgYN575SGMiJbe6ZwIwULtlC/7SedDyayBdX/UMBjkfKvyvx9tUdsfeUMWwhLX2
YTx3/Ma1IhdwBIvbIGBL
-----END CERTIFICATE-----
-----BEGIN CERTIFICATE-----
MIIBkTCCARgCAQEwCgYIKoZIzj0EAwMwRTELMAkGA1UEBhMCQVUxEzARBgNVBAgM
ClNvbWUtU3RhdGUxITAfBgNVBAoMGEludGVybmV0IFdpZGdpdHMgUHR5IEx0ZDAe
Fw0yNDA3MDQxMTI0NDZaFw0yNTA3MDQxMTI0NDZaMEUxCzAJBgNVBAYTAkFVMRMw
EQYDVQQIDApTb21lLVN0YXRlMSEwHwYDVQQKDBhJbnRlcm5ldCBXaWRnaXRzIFB0
eSBMdGQwWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAASdAgJZTIzvmeLYyWEPAt6D
xsALWO8iLk54JMA3UwNi72lAvTw6VVPiH6K5lESlPAoOayge4oYEp6v8s409J5Z/
MAoGCCqGSM49BAMDA2cAMGQCMGT+Gr+sibbbUI5TRhhMcg+UgSAPVizAjmxpWdtj
RVSsv+6Gh+LxvwqCoQW8vWJOPQIwK+08D7up/Ysguy5v8z+iTrfNhjUnqie0z1LQ
Ryep+3L/W0aw+ySuSI0paUxeNN5K
-----END CERTIFICATE-----
```
It is successfully parsed (but second certificate catch error in this place TTLS_ERR_PEM_FEATURE_UNAVAILABLE but `r` will be overwritten for third certificate .).
If we remove third certificate we return error, because it is a last cert in chain and r will be TTLS_ERR_PEM_FEATURE_UNAVAILABLE. 
If first certificate is not parsed but second parsed in fact we don't parse our certificate, parse CA  certificate and work as everything is ok!

Reply:
Just add `Proc-Type: 4,ENCRYPTED` in you certificate instead of first 22 bytes to simulate it

Reply:
This is mbed TLS inherited logic: if we observe any error on certificate parsing, but we see the certificate boundaries (header and footer), we just skip it and continue with the next certificate. I think we should terminate parsing if any of the certificates is invalid - a client won't validate such certificate chain either

Reply:
Fixed, now I see
```
[  874.310593] [tempesta tls] Warning: Error 1400 on parsing certificate with id 1 in the chain
[  874.310595] [tempesta fw] ERROR: tls_certificate: Invalid certificate specified, err=1400
[  874.314550] [tempesta fw] ERROR: configuration parsing error:
[  874.314550]   12: tls_certificate /root/tempesta/etc/tfw-root-test.crt;
[  874.314550]       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
in dmesg on loading the certificate from https://github.com/tempesta-tech/tempesta/pull/2157#discussion_r1665598468